library/Businessprocess/BpNode.php

```diff
@@ -460,6 +462,16 @@ public function reCalculateState()
             case self::OP_OR:
                 $sort_state = min($sort_states);
                 break;
+            case self::OP_DEGRADED:
+                $maxState = max($sort_states);
+                $maxIcingaState = $maxState >> self::SHIFT_FLAGS;
```

Comment:
This looks like a candidate for `$this->sortStateTostate($maxState)` to me. But you're comparing `$maxIcingaState` below with values in `$this->stateToSortStateMap` which then wouldn't work anymore... I'm currently not sure what's wrong or right here, but it *feels* wrong, just by looking at how `self::SHIFT_FLAGS` is used elsewhere.

I believe you should use `$maxIcingaState = $this->sortStateTostate($maxState);` and in the comparison below `$maxIcingaState === self::ICINGA_CRITICAL`.

Reply:
Hi @nilmerg, we tried to apply the suggested changes. It seems correct from a logical point of view. Unfortunately, the `getSortingState` function calls the `reCalculateState` one while calling the `isInDowntime`. As you can imagine the result is an infinite loop and we get the following error:
```
Uncaught Error: Maximum function nesting level of '256' reached, aborting! in /usr/share/icingaweb2/modules/businessprocess/library/Businessprocess/BpNode.php:424
```

Reply:
That's correct for `$this->getSortingState()`, but here I suggested to use `$this->sortStateTostate()` which isn't affected by this.